pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+	getPrometheusPods(ctx context.Context) (map[int32][]*corev1.Pod, error)
+
+	getPodMonitorings(ctx context.Context) ([]monitoringv1.PodMonitoring, error)
+	getClusterPodMonitorings(ctx context.Context) ([]monitoringv1.ClusterPodMonitoring, error)
+
+	patchPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.PodMonitoring) error
+	patchClusterPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.ClusterPodMonitoring) error
+
+	// Returns the target fetcher, with an zero-index number `thread` that signifies which goroutine this is for.
+	getTargetFetcher(ctx context.Context, thread uint16) (targetFetcher, error)
+}
+
+// Responsible for fetching the targets given a pod. One is created per goroutine.
+type targetFetcher interface {
+	getTarget(ctx context.Context, logger logr.Logger, port int32, pod *corev1.Pod) (*prometheusv1.TargetsResult, error)
+}
+
+// targetStatusReconciler to hold cached client state and source channel.
+type targetStatusReconciler struct {
+	ch            chan event.GenericEvent
+	opts          Options
+	manager       dataManager
+	tickerBuilder func() Ticker
+	logger        logr.Logger
+}
+
+// newTargetStatusReconciler creates a new targetStatusReconciler and kicks off
+// the reconcile loop by sending the first event to the source channel.
+func NewTargetStatusPoller(logger logr.Logger, restConfig *rest.Config, manager manager.Manager, registry prometheus.Registerer, opts Options) (TargetStatusProcessor, error) {
+	dataManagerBuilder, err := newControllerDataManagerBuilder(restConfig, opts, manager.GetClient())
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler, err := newTargetStatusPoller(logger, opts, registry, dataManagerBuilder)
+	err = ctrl.NewControllerManagedBy(manager).
+		Named("target-status").
+		For(
+			// We don't actually listen to any DaemonSet, just the channel.
+			&appsv1.DaemonSet{},
+			builder.WithPredicates(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				return false
+			})),
+		).
+		Watches(&source.Channel{
+			Source: reconciler.ch,
+		}, &handler.EnqueueRequestForObject{}).
+		Complete(reconciler)
+	if err != nil {
+		return nil, errors.Wrap(err, "create target status controller")
+	}
+
+	return reconciler, nil
+}
+
+func newTargetStatusPoller(logger logr.Logger, opts Options, registry prometheus.Registerer, builder dataManagerBuilder) (*targetStatusReconciler, error) {
+	if err := registry.Register(targetStatusDuration); err != nil {
+		return nil, err
+	}
+
+	ch := make(chan event.GenericEvent, 1)
+
+	dataManager, err := builder.build()
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler := &targetStatusReconciler{
+		ch:      ch,
+		opts:    opts,
+		manager: dataManager,
+		logger:  logger,
+	}
+
+	return reconciler, nil
+}
+
+type emptyObject struct{}
+
+func (r *targetStatusReconciler) Start(ctx context.Context, tickerBuilder func() Ticker) error {
+	r.tickerBuilder = tickerBuilder
```

Comment:
This would definitely work for managed collections but for the self-hosted option (or even developers), users can create, delete and re-create `DaemonSet`, causing the reconciler to retrigger and then you'd have two reconciler loops.

An alternative idea I had to remove the `Start` is to use `manager.Add` to add a callback when the manager is started and add an event into the queue. This ensures the controller is started only once, and started at the same time as the manager without needing to call a `Start` method from the outside.

Thoughts?

Reply:
The consequence is we'd be having double the amount of threads polling targets at a time. This could potentially happen in a deployed 

I thought of a few ways we can handle this:
1. Don't.
2. Use a fake object in the `For` in conjunction with themanager.Add`.
3. Use a `DaemonSet` and creation filter in the `For` in conjunction with an `EventHandler` to only accept the first ever create event.
4. If `controller-runtime` gave us more control over event handling, but I couldn't figure this one out based on the Go docs.